### ISSUE TITLE
Intermittent timeout in /websockets/cookies/006.html

### ISSUE THREAD

**Issue by nox** [servo/servo#9803]
```
  ▶ TIMEOUT [expected OK] /websockets/cookies/006.html

  ▶ Unexpected subtest result in /websockets/cookies/006.html:
  │ TIMEOUT [expected PASS] WebSockets: setting Secure cookie with document.cookie, checking ws request
  └   → Test timed out
```

**Comment by Manishearth** [comment]
This is a fun one. `document.cookie` works by sending a `SetCookiesForUrl` message to the resource task. This will be at some point handled by an event loop which will dump the cookies to the cookie jar.

On the other hand, websockets do not participate in this event loop, and read from the cookie jar directly. The cookie jar is behind a mutex so this is still safe.

The existence of the two synchronization mechanisms breaks the happens-before (e.g. a setcookies message is queued but the websocket reads directly from the jar first)

**Comment by jdm** [comment]
The fix for this and #9772 is to undo the cookie-related parts of #7654 (HSTS is fine because only the network stack interacts with it).

**Comment by jdm** [comment]
Actually, I'm not sure why we still don't have happens-before if `document.cookie` is set before creating the WebSocket. The resource thread is in charge of processing both the `SetCookiesForUrl` and `WebscketConnect` messages, so the cookies would be updated before the connection is initiated...

**Comment by Manishearth** [comment]
Huh, right.

**Comment by bobthekingofegypt** [comment]
Is it ok if I try to take a look at these intermittent websocket issues?

There seems to be a few,  I think are probably all related.

**Comment by Manishearth** [comment]
Go ahead!

**Comment by bobthekingofegypt** [comment]
I had some time to have a look at this one yesterday, might not get a chance again till later in the week.

I had a question about the test itself; this test appears to have the code branches for when the SCHEME_DOMAIN_PORT is set to 'wss' and also when it's not.  As it's testing that secure cookies are sent on wss and not on ws.  Looking at the other tests they usually have a separate test for each case, is there a reason that this one contains both paths, does it somehow get run both ways?


tl;dr - I have found a possible cause, but not a fix.

What follows is a dump of my current thoughts.

I have found a couple of things so far, first one I think is a bug in Servo's handling of client initiated close events.  I don't know if this should maybe be a separate issue, or I may just be wrong it would be good if someone could verify, but...

There are two ways to cleanly disconnect the websocket, client initiated and server initiated.  Server initiated looks like it should work fine but client initiated will not.  This isn't terminal but results in unnecessary things being sent in the channel.

Steps in client close:
* javascript calls websocket.close()
* calls Close in components/script/dom/websocket.rs
* Close function sets ready state to closing
* Close function sends the WebSocketDomAction::Close event
* components/net/websocket_loader picks up the event and sends the close message down the socket
* pywebsocket correctly echos back the close
* components/net/websocket_loader picks up the echo'd back close
* identifies the opcode as Close and then incorrectly echo's the close back again
* pywebsocket huffs and sends back an http error

I think this may effect some of the other intermittent websocket tests that are listening for onclose/onerror but I haven't checked them yet.


Now the part that is stumping me a bit.  I have found a possible cause for the intermittent in this issue but I need to verify why that might have caused an issue.  Normally when you run this test everything is fine, it does a couple of things

* handshake
* echo cookie back
* send a close op to the client

Normally this will happen over a minimum of 4 separate TCP packets; websocket upgrade, text response, close, close(masked from client).
However when the test fails the wpt python sever does something a bit different, it decides to embed the text and close websocket packets inside one TCP packet.  This could be a red herring as it should work fine, each websocket packet is framed individually so should be parsable.  But it seems very co-incidental that it happens everytime there is a failure.  I guess the next step would be to checkout rust-websocket and see what is happening in their.  Does anyone else have any idea's why this might cause an issue, or seen something like it before? Or am I chasing a ghost by following this path?

**Comment by bobthekingofegypt** [comment]
I've been trying to recreate this issue outside of Servo, I've created a repo with a test case that seems to trigger lock-ups on my machine.  Could someone please check it out and give it a go.  Apologies for the convoluted process but I can't make it happen reliably.

https://github.com/bobthekingofegypt/servo-cookie-inter

It may take a while to trigger, usually takes a few minutes on my machine.  If you are running a network monitor can you check the last few websocket packets sent and report back.


To confuse things a little more, I think I'm chasing a red herring on the cause.  What I was seeing was all successful communication looking like this, with the text frame and the close frame in separate tcp frames:
![success-separate-close](https://cloud.githubusercontent.com/assets/332141/13900050/107a7382-edf5-11e5-9fa3-a75580ac8c0f.png)

Then a failure is always preceded by this, where they are combined:
![failed-combined-close](https://cloud.githubusercontent.com/assets/332141/13900053/2ce0b536-edf5-11e5-8d61-fad8fb93d78d.png)

The problem is I dug into the code for pywebsockets and found it's possible for me to craft my own websocket frames and combine them before sending them down the socket, so I actually created another custom handler (contained inside the repo) that does exactly that.  This recreates the failure condition but has no negative effect on rust-websocket at all, everything works exactly as you would expect.  Which is ... annoying.

**Comment by jdm** [comment]
How long does it usually take for your program to lock up? I haven't reproduced any problem on OS X 10.9.5 yet.

**Comment by jdm** [comment]
On the other hand it locks up immediately in my ubuntu VM (using `while true`; it locks up on the second iteration using `while sleep 1`).

**Comment by bobthekingofegypt** [comment]
I probably should have said, I run linux. kernel 4.1.12-gentoo to be more specific.  Do you know what OS the build server intermittent's are happening on?

As for triggering times; I can sometimes run it for 2-3 minutes before it fails, sometimes it fails straight away.

**Comment by jdm** [comment]
The intermittent failures on build machines occur in Ubuntu Trusty.

**Comment by bobthekingofegypt** [comment]
I didn't realise someone else was going to be looking into this otherwise I would have said this last night, but I believe I understand why this is happening - I just haven't had time yet to verify it.

Worth noting is that this problem is not linux only, it happens on OSX as well, it really depends how fast your machine is executing at the time, I guess the slowness of the VM helps trigger it.
 
Anyway I finally stopped trying to make a single shot reproducible test case and instead dug through the code to work out in my head how it might be going wrong.  

What I believe is happening is a problem with rust-websocket (I will raise a bug there once I have a chance to formulate my thoughts).

If you break it down into events what happens in the test program and in the wpt tests is:
- program thread is created, this thread is responsible for creating the connection and processing/validating the handshake response.
- once validated then a receiving thread is initialised that blocks waiting for a response and a sending thread is initialised that waits for something to send.
- websocket server sends the cookies (or "none" in the test) back to the client who reads it and the close request entirely in rust-websocket land
- close response is echoed back and everyone is happy.

When it goes wrong I believe what is happening is:
- program thread is created, this thread is responsible for creating the connection and processing/validating the handshake response.
- sends the GET request waiting for the the upgrade response
- consumes the response, but this is inside hyper so it doesn't pay attention to websocket framing - it just consumes the entire socket but only cares about the headers and ignores the body of the frame.  This is a problem depending on timing as the socket may already contain the 2 websocket packets.
- the send and receive threads startup, but the actual websocket packets have already been consumed so everyone just sits waiting for something to happen and nothing ever does.

I hope to progress this tomorrow or Easter Monday when I can sit at my computer.

**Comment by perlun** [comment]
> I didn't realise someone else was going to be looking into this otherwise I would have said this last night, but I believe I understand why this is happening - I just haven't had time yet to verify it.

No worries whatsoever Bob - I was just looking for areas where I could contribute. It seems like you are quite far in sorting this out yourself.

> I hope to progress this tomorrow or Easter Monday when I can sit at my computer.

I'll leave it for the time being. Keep us posted; it would be soooo great to get this bug fixed. :smiley:

**Comment by bobthekingofegypt** [comment]
@jdm I submitted a patch to rust-websocket, it was accepted and merged and a new version released (the issue is still marked as open as can be seen in the link above but the pull request is done).

Can we bump rust-websocket upto 0.16.1 in servo?

**Comment by jdm** [comment]
@bobthekingofegypt Absolutely! Use `./mach cargo-update -p websocket` if you'd like to do so.

**Comment by nox** [comment]
https://github.com/servo/servo/pull/10240

**Comment by bobthekingofegypt** [comment]
@jdm will take a look at doing that later.

With the other issue I found earlier, incorrect client close handling (https://github.com/servo/servo/issues/9803#issuecomment-196424406).  Do you have an idea for how that should be solved?  Should a guard be replicated on the websocket_loader side of the channel that replicates the behaviour from the script/dom/websocket side?  Or should the ready_state be shared somehow between the two?

**Comment by nox** [comment]
@bobthekingofegypt I did a PR, see my previous comment. Nice work finding that fix btw.

**Comment by jdm** [comment]
@bobthekingofegypt Perhaps we should just store a boolean indicating that we're waiting to see a Close opcode from the server and avoid echoing it if that's the case.

**Comment by bobthekingofegypt** [comment]
@nox cheers.

@jdm Sounds good.

**Comment by bobthekingofegypt** [comment]
@jdm well I remember, I asked this previously but I'm still unsure about this test case itself.

> I had a question about the test itself; this test appears to have the code branches for when the SCHEME_DOMAIN_PORT is set to 'wss' and also when it's not. As it's testing that secure cookies are sent on wss and not on ws. Looking at the other tests they usually have a separate test for each case, is there a reason that this one contains both paths, does it somehow get run both ways?

The constants.js file for the websocket tests also contains the fixme
```
//FIXME: Add support for wss
```
Which suggests wss isn't supported yet.

**Comment by jdm** [comment]
I agree, that's pretty weird. It should be filed upstream in https://github.com/w3c/web-platform-tests/ though, and cc zcorpan.

**Comment by mrobinson** [comment]
This test is no longer disabled, so likely isn't intermittent any longer.
